unixfs/src/file/adder.rs

```diff
@@ -0,0 +1,723 @@
+        }
+    }
+}
+
+/// BalancedCollector creates balanced UnixFs trees, most optimized for random access to different
+/// parts of the file. Currently supports only link count threshold or the branching factor.
+#[derive(Clone)]
+pub struct BalancedCollector {
+    branching_factor: usize,
+    // reused between link block generation
+    reused_links: Vec<PBLink<'static>>,
+    // reused between link block generation
+    reused_blocksizes: Vec<u64>,
+}
+
+impl fmt::Debug for BalancedCollector {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            fmt,
+            "BalancedCollector {{ branching_factor: {} }}",
+            self.branching_factor
+        )
+    }
+}
+
+impl std::default::Default for BalancedCollector {
+    fn default() -> Self {
+        Self::with_branching_factor(174)
+    }
+}
+
+impl BalancedCollector {
+    /// Configure Balanced collector with the given branching factor.
+    pub fn with_branching_factor(branching_factor: usize) -> Self {
+        assert!(branching_factor > 0);
+
+        Self {
+            branching_factor,
+            reused_links: Vec::new(),
+            reused_blocksizes: Vec::new(),
+        }
+    }
+
+    fn flush_links(&mut self, pending: &mut Vec<Link>, finishing: bool) -> Vec<(Cid, Vec<u8>)> {
+        /*
+
+        file    |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -|
+        links#0 |-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
+        links#1 |-------|-------|-------|-------|-------|-------|-------|\   /
+        links#2 |-------------------------------|                         ^^^
+                                                                    one short
+
+        #flush_buffered_links(...) first iterations:
+
+        file    |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -|
+        links#0 |-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
+        links#1 |-------|-------|-------|-------|-------|-------|-------|==1==|
+        links#2 |-------------------------------|==========================2==|
+
+        new blocks #1 and #2
+
+        #flush_buffered_links(...) last iteration:
+
+        file    |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -|
+        links#0 |-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
+        links#1 |-------|-------|-------|-------|-------|-------|-------|--1--|
+        links#2 |-------------------------------|--------------------------2--|
+        links#3 |==========================================================3==|
+
+        new block #3 (the root block)
+        */
+
+        let mut ret = Vec::new();
+
+        let mut reused_links = std::mem::take(&mut self.reused_links);
+        let mut reused_blocksizes = std::mem::take(&mut self.reused_blocksizes);
+
+        if let Some(need) = self.branching_factor.checked_sub(reused_links.capacity()) {
```

Comment:
don't know how precisely the capacity accounting can go here, as we're not using `reserve_exact`

Reply:
I wished there was a `ensure_capacity(x)` but I may have misunderstood something here. the idea is to delay `Vec::with_capacity(self.branching_factor)` from the builder time to first use.

Reply:
the worst case scenario is just a few extra allocations, so definitely not a blocker

Reply:
I can't see how `reserve_exact` would be better. I guess we dont mind if this becomes longer, but it should not become.

I wonder if this could be an stack vec instead.. Probably a bit overboard but I'll try that while waiting.

Reply:
Didn't test but this might be possible...